### PR TITLE
Pytorch model Op fusion fix

### PR DESCRIPTION
**Pull request by KavithaTipturMadhu**
NumUses check is removed because a different alloc could be used as a result and a different operand could be used as an operand, as demonstrated in the pytorch example.

### CODE REVIEW

File: test/Passes/xsmm-combine.mlir [hunk 1]
```diff
@@ -262,3 +262,112 @@ func.func @none_on_binary_add_bf16(%arg0: memref<256x128xbf16>) -> memref<256x51
...
+// CHECK: scf.forall (%[[ARG1:.*]], %[[ARG2:.*]]) in (8, 32) {
+// CHECK:  %[[alloc_3:.*]] = memref.alloc() {alignment = 64 : i64} : memref<32x32xbf16>
+// CHECK:  %[[temp6:.*]] = xsmm.unary.dispatch zero [32, 32, 1, 32] flags = (bcast_scalar) data_type = bf16
+// CHECK:  xsmm.unary zero(data_type = bf16, %[[temp6]], %[[cst]], %[[alloc_3]]) : (i64, bf16, memref<32x32xbf16>) -> ()
+// CHECK:  %[[temp7:.*]] = xsmm.fused_brgemm.dispatch [32, 32, 32, 32, 32, 32, 1024, 1024][add,relu]  flags = (vnni_b)  binary_flags = (bcast_col_in0)  unary_flags = (none) data_type = bf16
+// CHECK:  xsmm.fused_brgemm(data_type = bf16, %[[temp7]], %{{.*}}, %{{.*}}, %[[alloc_3]], %{{.*}}, %[[c32_i64]]) 
+// CHECK: scf.forall (%[[ARG1:.*]], %[[ARG2:.*]]) in (8, 32) {
+// CHECK:  %[[alloc_3:.*]] = memref.alloc() {alignment = 64 : i64} : memref<32x32xbf16>
+// CHECK:  %[[temp6:.*]] = xsmm.unary.dispatch zero [32, 32, 1, 32] flags = (bcast_scalar) data_type = bf16
+// CHECK:  xsmm.unary zero(data_type = bf16, %[[temp6]], %[[cst]], %[[alloc_3]]) : (i64, bf16, memref<32x32xbf16>) -> ()
```
Comment (rengolin) [inline test/Passes/xsmm-combine.mlir:365]: Why is the zero not being fused as beta=0?

**Review by rengolin — COMMENTED** [review]
I agree this isn't scalable. I'd first try to remove the alloc by other means, which could very well be added to the upstream cleanups that @matthias-springer has added already.

File: test/Passes/xsmm-combine.mlir (hunk 1 above)
Comment (KavithaTipturMadhu) [reply test/Passes/xsmm-combine.mlir:365]: We don't fuse zeros in our current implementation yet, just brgemm, add and relu. We will need to extend the implementation for zeros. I'll update the patch with that feature.

Comment (adam-smnk) [reply test/Passes/xsmm-combine.mlir:365]: A bit tangential but we also don't fuse brgemm with add without relu?

File: lib/TPP/Dialect/Xsmm/XsmmUtils.cpp [hunk 2]
```diff
@@ -309,13 +312,13 @@ FailureOr<FusedMatch> getFusedBrgemmSequenceFromProducer(Operation *op) {
...
     }
 
-    // Make sure this is a chain, ie. at least once in inputs and outputs
-    int numUses = std::count(user->getOperands().begin(),
-                             user->getOperands().end(), op->getResult(0));
     // At least one input and the last operand (output) is the same buffer
-    if (numUses < 2 ||
-        user->getOperands()[user->getOperands().size() - 1] != op->getResult(0))
+    if (user->getOperands()[user->getOperands().size() - 1] !=
+            op->getResult(0) &&
```
Comment (adam-smnk) [inline lib/TPP/Dialect/Xsmm/XsmmUtils.cpp:317]: Shouldn't this be OR?

**Comment by adam-smnk** [comment]
I can't exactly put my finger on it but the conditions probably need more testing.

Here's one regression, I think this shouldn't fuse and it doesn't on main:
```
memref.global "private" constant @buf : memref<4x32x32xf32> = dense<1.601560e+00> {alignment = 64 : i64}
memref.global "private" constant @buf1 : memref<32xf32> = dense<1.601560e+00> {alignment = 64 : i64}

func.func @entry(%alloc_0: memref<8x8x32x32xf32>, %alloc: memref<8x4x32x32xf32>, %alloc1: memref<32x32xf32>) {
  %1 = memref.get_global @buf1 : memref<32xf32>
  %2 = memref.get_global @buf : memref<4x32x32xf32>
  %c4_i64 = arith.constant 4 : i64
  %cst = arith.constant 0.000000e+00 : f32

  scf.forall (%arg1, %arg2) in (8, 8) {
    %subview = memref.subview %alloc_0[%arg1, %arg2, 0, 0] [1, 1, 32, 32] [1, 1, 1, 1] : memref<8x8x32x32xf32> to memref<32x32xf32, strided<[32, 1], offset: ?>>
    %3 = xsmm.unary.dispatch zero [32, 32, 1, 32] flags = (bcast_scalar) data_type = f32
    xsmm.unary zero(data_type = f32, %3, %cst, %subview) : (i64, f32, memref<32x32xf32, strided<[32, 1], offset: ?>>) -> ()
    %subview_3 = memref.subview %alloc[%arg1, 0, 0, 0] [1, 4, 32, 32] [1, 1, 1, 1] : memref<8x4x32x32xf32> to memref<4x32x32xf32, strided<[1024, 32, 1], offset: ?>>
    %4 = xsmm.brgemm.dispatch [32, 32, 32, 32, 32, 32, 1024, 1024] flags = (none) data_type = f32
    xsmm.brgemm(data_type = f32, %4, %subview_3, %2, %subview, %c4_i64) : (i64, memref<4x32x32xf32, strided<[1024, 32, 1], offset: ?>>, memref<4x32x32xf32>, memref<32x32xf32, strided<[32, 1], offset: ?>>, i64) -> ()
    %5 = xsmm.binary.dispatch add [32, 32, 32, 32, 32] flags = (bcast_col_in0) data_type = f32
    xsmm.binary add(data_type = f32, %5, %1, %subview, %alloc1) : (i64, memref<32xf32>, memref<32x32xf32, strided<[32, 1], offset: ?>>, memref<32x32xf32>) -> ()
    %6 = xsmm.unary.dispatch relu [32, 32, 32, 32] flags = (none) data_type = f32
    xsmm.unary relu(data_type = f32, %6, %subview, %subview) : (i64, memref<32x32xf32, strided<[32, 1], offset: ?>>, memref<32x32xf32, strided<[32, 1], offset: ?>>) -> ()
  }

  return
}
```
Note `%alloc1` in `xsmm.binary add` before relu.

File: lib/TPP/Dialect/Xsmm/XsmmUtils.cpp (hunk 2 above)
Comment (KavithaTipturMadhu) [reply lib/TPP/Dialect/Xsmm/XsmmUtils.cpp:317]: No it should be and because the result should match either the last or the penultimate operand.

**Comment by KavithaTipturMadhu** [comment]
> I can't exactly put my finger on it but the conditions probably need more testing.
> 
> Here's one regression, I think this shouldn't fuse and it doesn't on main:
> 
> ```
> memref.global "private" constant @buf : memref<4x32x32xf32> = dense<1.601560e+00> {alignment = 64 : i64}
> memref.global "private" constant @buf1 : memref<32xf32> = dense<1.601560e+00> {alignment = 64 : i64}
> 
> func.func @entry(%alloc_0: memref<8x8x32x32xf32>, %alloc: memref<8x4x32x32xf32>, %alloc1: memref<32x32xf32>) {
>   %1 = memref.get_global @buf1 : memref<32xf32>
>   %2 = memref.get_global @buf : memref<4x32x32xf32>
>   %c4_i64 = arith.constant 4 : i64
>   %cst = arith.constant 0.000000e+00 : f32
> 
>   scf.forall (%arg1, %arg2) in (8, 8) {
>     %subview = memref.subview %alloc_0[%arg1, %arg2, 0, 0] [1, 1, 32, 32] [1, 1, 1, 1] : memref<8x8x32x32xf32> to memref<32x32xf32, strided<[32, 1], offset: ?>>
>     %3 = xsmm.unary.dispatch zero [32, 32, 1, 32] flags = (bcast_scalar) data_type = f32
>     xsmm.unary zero(data_type = f32, %3, %cst, %subview) : (i64, f32, memref<32x32xf32, strided<[32, 1], offset: ?>>) -> ()
>     %subview_3 = memref.subview %alloc[%arg1, 0, 0, 0] [1, 4, 32, 32] [1, 1, 1, 1] : memref<8x4x32x32xf32> to memref<4x32x32xf32, strided<[1024, 32, 1], offset: ?>>
>     %4 = xsmm.brgemm.dispatch [32, 32, 32, 32, 32, 32, 1024, 1024] flags = (none) data_type = f32
>     xsmm.brgemm(data_type = f32, %4, %subview_3, %2, %subview, %c4_i64) : (i64, memref<4x32x32xf32, strided<[1024, 32, 1], offset: ?>>, memref<4x32x32xf32>, memref<32x32xf32, strided<[32, 1], offset: ?>>, i64) -> ()
>     %5 = xsmm.binary.dispatch add [32, 32, 32, 32, 32] flags = (bcast_col_in0) data_type = f32
>     xsmm.binary add(data_type = f32, %5, %1, %subview, %alloc1) : (i64, memref<32xf32>, memref<32x32xf32, strided<[32, 1], offset: ?>>, memref<32x32xf32>) -> ()
>     %6 = xsmm.unary.dispatch relu [32, 32, 32, 32] flags = (none) data_type = f32
>     xsmm.unary relu(data_type = f32, %6, %subview, %subview) : (i64, memref<32x32xf32, strided<[32, 1], offset: ?>>, memref<32x32xf32, strided<[32, 1], offset: ?>>) -> ()
>   }
> 
>   return
> }
> ```
> 
> Note `%alloc1` in `xsmm.binary add` before relu.

You're right, I agree that the conditions need more testing and probably more automated tests too. The test you've mentioned however, is not fusing in main because of numUses check and not because of IO chaining like it should be. I will update the patch and add more test cases. In the meantime, please do let me know if you can think of any more edge cases I may have missed.

File: lib/TPP/Dialect/Xsmm/XsmmUtils.cpp (hunk 2 above)
Comment (adam-smnk) [reply lib/TPP/Dialect/Xsmm/XsmmUtils.cpp:317]: Alright, I see that now. Thanks.

**Comment by KavithaTipturMadhu** [comment]
Without fusion: 
gemm_fp32_torch             :   114.583 gflops
gemm_bf16_torch             :   802.782 gflops
mlp_fp32_torch              :   112.595 gflops
mlp_bf16_torch              :   716.085 gflops
 
With fusion:
gemm_fp32_torch             :   114.604 gflops
gemm_bf16_torch             :   800.157 gflops
mlp_fp32_torch              :   114.853 gflops
mlp_bf16_torch              :   768.769 gflops

File: test/Passes/xsmm-combine.mlir (hunk 1 above)
Comment (KavithaTipturMadhu) [reply test/Passes/xsmm-combine.mlir]: Will address both these issues in a different PR as they are in different files.

File: benchmarks/mlir/pytorch/torch-dynamo-mlp-bf16-3x1024.mlir [hunk 3]
```diff
@@ -17,51 +17,39 @@ module attributes {torch.debug_module_name = "_lambda"} {
     %cst_4 = arith.constant dense<1.6> : tensor<1024x1024xbf16>
     %cst_5 = arith.constant 0.000000e+00 : bf16
     %0 = tensor.empty() : tensor<1024x1024xbf16>
-    %1 = linalg.generic {indexing_maps = [#map, #map1], iterator_types = ["parallel", "parallel"]} ins(%cst_4 : tensor<1024x1024xbf16>) outs(%0 : tensor<1024x1024xbf16>) {
```
Comment (rengolin) [inline benchmarks/mlir/pytorch/torch-dynamo-mlp-bf16-3x1024.mlir:20]: We should not change the benchmark IR, since this was as close as possible form what we get from PyTorch. If we change, then we're not sure we cna handle PyTorch anymore.

File: lib/TPP/Transforms/LinalgConvertCompareSelectToMaximumfPass.cpp [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+namespace tpp {
+
+struct LinalgConvertCompareSelectToMaximumf
+    : public OpRewritePattern<linalg::GenericOp> {
+  using OpRewritePattern<linalg::GenericOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(linalg::GenericOp op,
+                                PatternRewriter &rewriter) const override {
+
+    if (op.getBody()->getOperations().size() != 3 ||
```
Comment (rengolin) [inline lib/TPP/Transforms/LinalgConvertCompareSelectToMaximumfPass.cpp:37]: You can simplify this check + get with a single patter:
```c++
  auto comp = dyn_cast<...>(...);
  if (!comp)
    return failure();

  auto sel = dyn_cast<...>(...);
  if (!sel)
    return failure();

  ...
```

File: lib/TPP/Transforms/LinalgConvertCompareSelectToMaximumfPass.cpp [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+        dyn_cast<arith::CmpFOp>(op.getBody()->getOperations().begin())
+            ->getOperands());
+    dyn_cast<YieldOp>(op.getBody()->getTerminator()).setOperand(0, maxf);
+    op.getOutputsMutable().clear();
+    ValueRange range{op.getInputsMutable()};
+    op.getOutputsMutable().append(range);
+    op.getInputsMutable().clear();
+    op.setIndexingMapsAttr(
+        ArrayAttr::get(rewriter.getContext(), op.getIndexingMaps()[0]));
+    op.getBody()->eraseArgument(1);
```
Comment (rengolin) [inline lib/TPP/Transforms/LinalgConvertCompareSelectToMaximumfPass.cpp:59]: You shouldn't really need to touch the generic op at all, just replace the arith ops inside.

File: benchmarks/mlir/pytorch/torch-dynamo-mlp-bf16-3x1024.mlir (hunk 3 above)
Comment (KavithaTipturMadhu) [reply benchmarks/mlir/pytorch/torch-dynamo-mlp-bf16-3x1024.mlir:20]: The one change absolutely required is the in-place addition, which we have to introduce because it adds a buffer otherwise. I'll revert the rest of the changes.

**Comment by KavithaTipturMadhu** [comment]
@rengolin I have updated the patch to check if the two add inputs are the same, to skip the test that we discussed.

File: lib/TPP/Transforms/ConvertAddInplacePass.cpp [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    if (op.getInputs()[0] == op.getInputs()[1])
+      return failure();
+    SmallVector<AffineMap> indexingMaps;
+    SmallVector<utils::IteratorType> iteratorTypes;
+    for (auto iteratorTypesArray : op.getIteratorTypesArray()) {
+      iteratorTypes.push_back(iteratorTypesArray);
+    }
+
+    Value inputs, outputs;
+    if (op.getIndexingMapsArray()[1] ==
```
Comment (rengolin) [inline lib/TPP/Transforms/ConvertAddInplacePass.cpp:52]: nit: perhaps a comment on how this is changing

File: lib/TPP/Transforms/LinalgConvertCompareSelectToMaximumfPass.cpp [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+
+    if (op.getBody()->getOperations().size() != 3)
+      return failure();
+    auto cmpf = dyn_cast<arith::CmpFOp>(&op.getBody()->getOperations().front());
+    if (!cmpf)
+      return failure();
+    auto select = dyn_cast<arith::SelectOp>(
+        std::next(op.getBody()->getOperations().begin(), 1));
+    if (!select)
+      return failure();
```
Comment (rengolin) [inline lib/TPP/Transforms/LinalgConvertCompareSelectToMaximumfPass.cpp:45]: I'd also add a check to make sure the `cmp` is with `zero` and the `select` is between the result of `cmp` and `zero`.

Two simple tests not matching:
* comparison / select with non-zero
* comparison / select with zero but comparison from an external (captured) value, not the result of `cmp`

**Review by rengolin — APPROVED** [review]
LGTM, thanks!
